File: Design-Patterns-Using-Swift/Singleton/SingletonDemo.swift
class Database {
    init() {
        print("Database created")
    }

    func addData(_ data: String) {
        print(data)
    }
}

class Singleton {
    private static var instance: Singleton?
    private var db: Database

    private init() {
        db = Database()
    }

    static func getInstance() -> Singleton {
        if instance == nil {
            instance = Singleton()
        }
        return instance!
    }

    func addData(_ data: String) {
        db.addData(data)
    }
}

// Client Code
let s1 = Singleton.getInstance()
let s2 = Singleton.getInstance()

print(ObjectIdentifier(s1))
print(ObjectIdentifier(s2))

s2.addData("Hello, world!")

/*
Database created
ObjectIdentifier(0x00005620d9423040)
ObjectIdentifier(0x00005620d9423040)
Hello, world!
*/

File: Design-Patterns-Using-Cpp/Singleton/SingletonDemo.cpp
#include <iostream>

class Database {
public:
    Database() {
        std::cout << "Database created" << std::endl;
    }

    void addData(const std::string& data) {
        std::cout << data << std::endl;
    }
};

class Singleton {
private:
    static Singleton* instance;
    Database db;

    Singleton() {
        instance = this;
    }

public:
    static Singleton* getInstance() {
        if (instance == nullptr) {
            instance = new Singleton();
        }
        return instance;
    }

    void addData(const std::string& data) {
        db.addData(data);
    }
};

Singleton* Singleton::instance = nullptr;

// Client code
int main() {
    Singleton* s1 = Singleton::getInstance();
    Singleton* s2 = Singleton::getInstance();

    std::cout << s1 << std::endl;
    std::cout << s2 << std::endl;

    s2->addData("Hello, world!");

    delete s1;  // Clean up allocated instance

    return 0;
}

/*
Database created
0x55ad35ddeeb0
0x55ad35ddeeb0
Hello, world!
*/